_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,263 @@
+
+```java
+@Component
+class UserRemovedListener {
+
+	...
+	@EventListener(condition = "#event.name eq 'reflectoring'")
+	void handleConditionalListener(UserRemovedEvent event) {
+		System.out.println(String.format("User removed (Conditional): %s", event.getName()));
+	}
+}
+```
+
+In the above example, the listener will be triggered with `UserRemovedEvent` only when the `#event.name` has value `'reflectoring'`,
+
+### Implementing `ApplicationListener`
+
+```java
+@Component
+class UserCreatedListener implements ApplicationListener<UserCreatedEvent> {
+
+	@Override
+	public void onApplicationEvent(UserCreatedEvent event) {
+		System.out.println(String.format("User created: %s", event.getName()));
+	}
+}
+```
+
+In the above example, we have created a listener by implementing `ApplicationLister` and the generic represents the type of event you want to listen. It is now possible to define your `ApplicationListener` implementation with nested generics information in the event type. When dispatching an event, the signature of your listener is used to determine if it matches said incoming event.
+
+### Asynchronous Event Listeners
+
+**By default spring events are synchronous, meaning the publisher thread blocks until all listeners have finished processing the event.**
+
+To make an event listener run in async mode, all we have to do is use the `@Async` annotation on that listener. To make the `@Async` annotation work, we also have to annotate one of our `@Configuration` classes or the `@SpringBootApplication` class with `@EnableAsync`.
+
+```java
+@Component
+class AsyncListener {
+
+	@Async
+	@EventListener
+	void handleAsyncEvent(String event) {
+		System.out.println(String.format("Async event recevied: %s", event));
+	}
+}
+```
+
+## Transaction Bound Events
+
+Spring allows us to bind an event listener to a phase of the current transaction. This allows events to be used with more flexibility when the outcome of the current transaction actually matters to the listener.
+
+The transaction module implements an `EventListenerFactory` that looks for the new `@TransactionalEventListener` annotation. So when you annotate your method with `@TransactionalEventListener` an extended event listener that is aware of the transaction is registered instead of the default.
+
+You can bind the listener to the following phases of transaction:  
+`AFTER_COMMIT` : Fire the event after the commit has completed successfully.  
+`AFTER_COMPLETION`: Fire the event after the transaction has completed.  
```

Comment:
Again, these sentences are copied straight from the JavaDocs. It's not wrong, but we're writing the article to add value, not to repeat the docs. At least add a sentence to each transaction phase that perhaps explains a use case where it makes sense to use.